### DISCUSSION TITLE
Getting Error Cannot invoke "io.opentelemetry.api.trace.Tracer.spanBuilder(String)" because "this.tracer" is null

### DISCUSSION THREAD

**Discussion by yatamraghu7** [open-telemetry/opentelemetry-java-instrumentation#12280] [Q&A] [answered]
Hi @trask and @laurit,

I am trying to create a new span using the trace Id that we are getting from web socket communication from another micro-service and getting below error

`Cannot invoke "io.opentelemetry.api.trace.Tracer.spanBuilder(String)" because "this.tracer" is null`

We are using the tracer as below in our code 

    private Tracer tracer = GlobalOpenTelemetry.getTracer("my-tracer");

Can someone tell us what could be the reason for this and below are the maven dependnecies
```

                 <dependency>
			<groupId>io.opentelemetry.instrumentation</groupId>
			<artifactId>opentelemetry-spring-boot-starter</artifactId>
			<version>2.7.0</version>
		</dependency>
		<dependency>
			<groupId>io.opentelemetry</groupId>
			<artifactId>opentelemetry-extension-trace-propagators</artifactId>
			<version>1.42.1</version>
		</dependency>
		<dependency>
			<groupId>io.opentelemetry</groupId>
			<artifactId>opentelemetry-api</artifactId>
			<version>1.42.1</version>
		</dependency>
```

We are trying to achieve the common traceId between the microservices which are communicating using web-socket.
It would be great if we can get some help at this topic.

![Screenshot from 2024-09-18 15-40-07](https://github.com/user-attachments/assets/02322b3b-dc82-4847-8779-02ce8b5f2c75)

Problem Can be reproduced using the code here in the repositroy

https://github.com/yatamraghu7/ManualInstrumentationOtel

Thanks

**Comment by laurit** [comment, marked as answer]
> We are using the tracer as below in our code
private Tracer tracer = GlobalOpenTelemetry.getTracer("my-tracer");

But you aren't https://github.com/yatamraghu7/ManualInstrumentationOtel/blob/d8272498d03dc1b2c81cabf8d705b48a415ff2dd/Producer/src/main/java/com/microservice/producer/Producer/Controller/ProducerController.java#L26-L27

**Reply by yatamraghu7** [reply]
Hi @laurit 

Thanks for the update. Do we need to add opentelemetry-sdk dependency to get the tracer ?
I have added that and this sdk is again looking for one more dependency i.e. opentelemetry-api-logs. If I don't add the api-logs I am getting below error 

Exception in thread "BatchLogRecordProcessor_WorkerThread-1" java.lang.NoClassDefFoundError: io/opentelemetry/api/incubator/logs/AnyValue

We wanted to avoid this extra dependencies if possible. Is there any way ?

Also as mentioned we are using  above provided dependencies which is using 1.42.1 i think the latest but the mvn dependency opentelemetry-api-logs comes with latest version as below

               <dependency>
			<groupId>io.opentelemetry</groupId>
			<artifactId>opentelemetry-api-logs</artifactId>
			<version>1.26.0-alpha</version>
		</dependency>

Our issue is that we are having dependency version mismatch as the 1.26.0-alpha is not compatible with 1.42.1 and for this I have to downgrade to 1.32.0 and with this I have achieved websocket manual tracing.

Can you please tell us that if we can do in a better way with our dependencies.

Thanks

**Reply by laurit** [reply]
You should not need extra dependencies besides `opentelemetry-spring-boot-starter` Have a look at https://opentelemetry.io/docs/zero-code/java/spring-boot-starter/getting-started/#dependency-management on how to use a bom to ensure that you get dependencies with correct versions.
Secondly when using spring boot starter there is no implicitly defined `Tracer` bean, try injecting `OpenTelemetry` instead. From the injected `OpenTelemetry` instance you can get a tracer.

**Reply by yatamraghu7** [reply]
Hi @laurit 

Thanks for the explanation. If we are sending a custom X-B3-TraceId and we wanted that traceId to be carried throughout the request flow and we dont want new traceId to be generated and I think injecting OpenTelemetry and getting a tracer from the instance wont change my traceId throughout the request flow right ?

Thanks

**Reply by laurit** [reply]
no, it won't.

**Comment by yatamraghu7** [comment]
Thanks @laurit for your support as always.
